plutus-metatheory/src/Algorithmic/Signature.lagda.md

```diff
@@ -0,0 +1,121 @@
+
+```
+module Algorithmic.Signature where
+```
+
+## Imports
+
+```
+open import Data.Nat using (suc)
+
+open import Relation.Binary.PropositionalEquality using (_≡_;refl;cong;sym)
+open Relation.Binary.PropositionalEquality.≡-Reasoning
+
+open import Function using (_∘_)
+
+open import Utils using (*;_∔_≣_)
+open import Type using (Ctx⋆;∅;_,⋆_;_⊢⋆_;Φ)
+open _⊢⋆_
+
+open import Type.BetaNormal using (_⊢Nf⋆_;_⊢Ne⋆_;renNf;embNf)
+open _⊢Nf⋆_
+open _⊢Ne⋆_
+
+import Type.RenamingSubstitution as ⋆
+open import Type.BetaNBE.Completeness using (reifyCR;idext;exte-lem)
+open import Type.BetaNBE.RenamingSubstitution 
+                         using (subNf;SubNf;renNf-subNf;subNf-cong;subNf-comp;subNf-cons;extsNf;subNf-lemma) 
+                         renaming (_[_]Nf to _[_])
+open import Type.BetaNBE.Stability using (stability)
+open import Builtin using (Builtin;signature)
+open import Builtin.Signature using (Sig;sig;nat2Ctx⋆;fin2∈⋆)
+open import Type.BetaNBE using (nf;reify;eval;idEnv;exte)
+open Builtin.Signature.FromSig Ctx⋆ (_⊢Nf⋆ *) nat2Ctx⋆ (λ x → ne (` (fin2∈⋆ x))) con _⇒_ Π 
+     using (sig2type;SigTy;sigTy2type)
+open SigTy
+
+
+```
+
+```
+btype : Builtin → Φ ⊢Nf⋆ *
+btype b = subNf (λ()) (sig2type (signature b))
+
+btype-∅ : ∀ {A : ∅ ⊢Nf⋆ *} → A ≡ subNf {∅} {∅} (λ()) {*} A
+btype-∅ {A} = begin
+             A
+            ≡⟨ sym (stability A) ⟩
+             nf (embNf A)
+           ≡⟨ cong nf (sym (⋆.sub-∅ (embNf A)  (embNf ∘  λ()))) ⟩
+             nf (⋆.sub (embNf ∘ λ()) (embNf A))
+           ≡⟨ refl ⟩
+             subNf (λ ()) A
+           ∎
+
+btype-sig2type : ∀ (b : Builtin) → btype {∅} b ≡ sig2type (signature b)
+btype-sig2type b = sym btype-∅
+
+btype-ren : ∀{Φ Ψ} b (ρ : ⋆.Ren Φ Ψ) → btype b ≡ renNf ρ (btype b)
+btype-ren b ρ = begin
+             btype b
+             ≡⟨ refl ⟩
+             subNf (λ()) (sig2type (signature b))
+             ≡⟨ subNf-cong {f = λ()} {renNf ρ ∘ λ ()} (λ ()) (sig2type (signature b)) ⟩
+             subNf (renNf ρ ∘ λ ()) (sig2type (signature b))
+             ≡⟨ renNf-subNf (λ()) ρ (sig2type (signature b)) ⟩
+             renNf ρ (btype b)
+           ∎
+
+btype-sub : ∀{Φ Ψ} b (ρ : SubNf Φ Ψ) → btype b ≡ subNf ρ (btype b)
+btype-sub b ρ = begin 
+           btype b
+          ≡⟨ refl ⟩
+           subNf (λ()) (sig2type (signature b))
+          ≡⟨ subNf-cong {f = λ()} {subNf ρ ∘ λ ()} (λ ()) (sig2type (signature b)) ⟩
+            subNf (subNf ρ ∘ (λ ())) (sig2type (signature b))
+          ≡⟨ subNf-comp (λ()) ρ (sig2type (signature b)) ⟩
+           subNf ρ (subNf (λ()) (sig2type (signature b)))
+          ≡⟨ refl ⟩
+           subNf ρ (btype b)
+          ∎
+```
+
+## Substitution in Signature types
+
+```
+subNf-Π : ∀{Φ Ψ J}(ρ : SubNf Φ Ψ)(B : Φ ,⋆ J ⊢Nf⋆ *) →  subNf ρ (Π B) ≡ Π (subNf (extsNf ρ) B)
+subNf-Π {Φ}{Ψ}{J} ρ B = begin
```

Comment:
this is clear but I worry may slow things down if agda has to execute it a lot

Reply:
Do you suggest to avoid all the uses of Eq-reasoning?

Reply:
Or just the extra `refl`s?

Reply:
I think it’s clear and nice to have. I think it should be ok to keep. If things slow down can strip things like this put.